tests/hypothesis_strategies/constraints.py

```diff
@@ -0,0 +1,252 @@
+"""Hypothesis strategies for constraints."""
+
+from functools import partial
+from typing import Any, Optional, Union
+
+import hypothesis.strategies as st
+from hypothesis import assume
+
+from baybe.constraints.conditions import (
+    SubSelectionCondition,
+    ThresholdCondition,
+    _valid_logic_combiners,
+)
+from baybe.constraints.continuous import (
+    ContinuousLinearEqualityConstraint,
+    ContinuousLinearInequalityConstraint,
+)
+from baybe.constraints.discrete import (
+    DiscreteDependenciesConstraint,
+    DiscreteExcludeConstraint,
+    DiscreteLinkedParametersConstraint,
+    DiscreteNoLabelDuplicatesConstraint,
+    DiscretePermutationInvarianceConstraint,
+    DiscreteProductConstraint,
+    DiscreteSumConstraint,
+)
+from baybe.parameters.base import DiscreteParameter
+from baybe.parameters.numerical import NumericalDiscreteParameter
+
+
+def sub_selection_conditions(superset: Optional[list[Any]] = None):
+    """Generate :class:`baybe.constraints.conditions.SubSelectionCondition`."""
+    if superset is None:
+        element_strategy = st.text()
+    else:
+        element_strategy = st.sampled_from(superset)
+    return st.builds(SubSelectionCondition, st.lists(element_strategy, unique=True))
+
+
+def threshold_conditions():
+    """Generate :class:`baybe.constraints.conditions.ThresholdCondition`."""
+    return st.builds(
+        ThresholdCondition, threshold=st.floats(allow_infinity=False, allow_nan=False)
+    )
+
+
+@st.composite
+def discrete_excludes_constraints(
+    draw: st.DrawFn, parameters: Optional[list[DiscreteParameter]] = None
+):
+    """Generate :class:`baybe.constraints.discrete.DiscreteExcludeConstraint`."""
+    if parameters is None:
+        parameter_names = draw(st.lists(st.text(), unique=True, min_size=1))
+        conditions = draw(
+            st.lists(
+                st.one_of(sub_selection_conditions(), threshold_conditions()),
+                min_size=len(parameter_names),
+                max_size=len(parameter_names),
+            )
+        )
+
+    else:
+        parameter_names = [p.name for p in parameters]
+
+        # Threshold conditions only make sense for numerical parameters
+        conditions = [
+            draw(st.one_of(sub_selection_conditions(p.values), threshold_conditions()))
+            if isinstance(p, NumericalDiscreteParameter)
+            else draw(sub_selection_conditions(p.values))
+            for p in parameters
+        ]
+
+    combiner = draw(st.sampled_from(list(_valid_logic_combiners)))
+    return DiscreteExcludeConstraint(parameter_names, conditions, combiner)
+
+
+@st.composite
+def discrete_dependencies_constraints(
```

Comment:
Hi @Scienfitz, sorry for the late review ... but here finally my (only two) comments:
* I see what you did here and imagine that it must have been quite confusing/tricky to get everything running because I recently attempted the same in the intervals strategy refactoring: you have one single flow of code that attempts to cover all four possible cases (i.e. arguments being None vs filled). All I can say at the moment: it makes it rather hard to follow the logic and be sure to not have forgotten any edge case... I realized this in the other PR and then changed gears and handled the cases separately, which resulted in significantly simpler code and was much easier to accomplish. That said, I don't urge you to change it but wanted to make you aware that the actual test of the logic is postponed anyway until we implement the search space strategies. Because that's where the created output distribution actually matters. So it will probably happen that you'll have to dig through this intertwined code logic again 🙃 For this PR, we don't actually need to care since all that is needed are serializable objects. 
* I think especially your `assume` will become rather inefficient in a real test later (-> searchspace strategies) and the logic can be also implemented without it. But again: you don't have to bother now if you want to but you have been warned :D

Other than that, I think we can finish the PR. Thoughts?

Reply:
makes sense, I dont think the improvements need to be implemented now, but in particular if you want to get rid of the assume already now then please do it. 

But considering these strats have to be changed anyway again once they should be combined with actual searchspaces, lets merge

Reply:
agree -> merge and bother later on :D